ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -73,6 +74,44 @@ where
     pub fn iter(&self) -> impl Iterator<Item = &([F; λ], [F; λ])> + Clone {
         self.uv_chunks.iter()
     }
+
+    /// This function allows to generate and set masks
+    ///
+    /// It outputs `(p_mask_from_left_prover,q_mask_from_right_prover)`.
+    ///
+    /// ## Errors
+    /// Errors when the length is too long such that masks cannot be set safely.
+    pub fn set_masks<C: Context>(
+        &mut self,
+        ctx: &C,
+        record_counter: &mut RecordId,
+    ) -> Result<(F, F), Error> {
+        if self.len() >= λ {
+            return Err(DZKPMasks);
+        }
+        // generate masks
+        // verifier on the right has p,
+        // therefore the right share is "implicitly sent" to the right ("communicated" via PRSS)
+        let (p_mask_from_left_prover, my_p_mask): (F, F) =
+            ctx.prss().generate_fields(*record_counter);
+        *record_counter += 1;
+        // and verifier on the left has q
+        // therefore the left share is "implicitly sent" to the left (communication via PRSS)
+        let (my_q_mask, q_mask_from_right_prover): (F, F) =
+            ctx.prss().generate_fields(*record_counter);
+        *record_counter += 1;
+
+        // compute final uv values
+        let (u_values, v_values) = &mut self.uv_chunks[0];
+        // shift first element to last position
+        u_values[SmallProofGenerator::RECURSION_FACTOR - 1] = u_values[0];
+        v_values[SmallProofGenerator::RECURSION_FACTOR - 1] = v_values[0];
```

Comment:
The position of the other values does not matter, it is just the mask that needs to in the 0 position (at least in some cases). It might be ok to put the mask in a different point but the current security analysis does not cover that.